### PR TITLE
feat: update toolbar placement and content spacing

### PR DESCRIPTION
**Pull request by austingreendev**
## Description

This PR includes:
- Additional tweaks on toolbar spacing
- Increased padding within the core editor and toolbar
- Updated relative spacing between editing content
- Added a PostCSS build process to source theme overrides from our `DEFAULT_THEME` instance

image

### Dev changes

- Moved from the `ClassicEditor` to the `DecoupledEditor`
  - Our "final" shared asset will be a CSS dist which includes CKEditor compatible theme overrides
  - This allows consumers to use the `DecoupledEditor` in a variety of experiences where the ClassicEditor isn't the best fit.
- Include a PostCSS build step that sources values from `react-theming`
  - When implementing our Garden simple vars I reserved the `__variables.css` file to only supply custom-properties provided by the CK theming interface.
  - Since we are providing a CSS dist that is consumed by another customized PostCSS config I am unable to use more common libraries like `postcss-env`

### Design

- The toolbar and editor padding/spacing have been increased to align with updated designs.
image
- Content spacing
  - All paragraph, list, codeblock, and other "block" elements have a default `margin-block-start/end` of `16px`
  - Heading 1's only receive a `margin-block-end`
  - All other headings are spaced equally
  - Horizontal lines have a block-start/end of 4px to align with the reduced heading margin
image
  - List items no longer receive padding, which aligns the line-heights with the default paragraph

### CODE REVIEW

File: README.md
Comment (austingreendev) [inline README.md:1]: Storybook mdx process is unable to handle these reference links. @jzempel is ok to use inline links to allow us to share this documentation with storybook?

File: .storybook/main.js
Comment (austingreendev) [inline .storybook/main.js:42]: This customization is only required for our local storybook environment. Consumers wouldn't have to worry about this step.

File: postcss.config.js
Comment (austingreendev) [inline postcss.config.js:46]: Due to limitations in [postcss-custom-properties](https://github.com/postcss/postcss-custom-properties), it is unable to replace CSS variables located in the `:root` selector.

Using `simple-vars` is a nice way to help differentiate between CK and Garden provided values.

File: src/theme/_variables.css
Comment (austingreendev) [inline src/theme/_variables.css:23]: `postcss-color-function` (and `postcss-env`) is blowing up when we attempt to use `color-mod` to add alpha to any value. There are syntax collisions in several CK theme files. Changing the order and removing cssnano minification didn't help in this case.

File: README.md
Comment (jzempel) [reply README.md:1]: Another option would be to keep the URLs simple and workaround storybook's bug like we did here: https://github.com/zendeskgarden/svg-icons/blob/main/src/.stories/README.stories.mdx 🤷

File: src/theme/_variables.css
Comment (jzempel) [reply src/theme/_variables.css]: `color-mod` has a particular no-comma syntax that can be tricky to deal with. Did you grep `css-components` for comparative usage? Also, as I recall, postcssrc needs the right stage and features set: https://github.com/zendeskgarden/css-components/blob/main/.postcssrc.js

Comment (austingreendev) [reply src/theme/_variables.css]: I was able to get `color-mod` working with the matching stage preset, but those conflicted with some of the features and syntax assumptions within the CK theme. I'm going to keep returning to this, but just wanted to call it out for now.

File: README.md
Comment (austingreendev) [reply README.md:1]: That is slick. Updated to use this logic.

File: src/theme/_variables.css
Comment (jzempel) [inline src/theme/_variables.css:23]: Did you attempt using [`rrggbbaa`](https://cssnext.github.io/features/#rrggbbaa-colors) with the `$ZD-PALETTE-BLUE-600` here?

File: src/theme/_variables.css
Comment (jzempel) [inline src/theme/_variables.css:48]: Use `shadowWidths.md` if possible.

File: src/theme/_variables.css
Comment (jzempel) [inline src/theme/_variables.css:57]: `rrggbbaa`?

**Review by jzempel — CHANGES_REQUESTED** [review]
@m-lai (cc @allisonacs) a few points for consideration:

- have you considered how the current HR treatment throws off vertical rhythm? A `<p>..<p>..<hr>..<p>..<p>` progression results in `16px..20px..20px..16px` of spacing. I'm going to keep asserting that 0 padding on the HR is the right amount (and provides a more general focus solution for "typearound" widgets – which would be very nice to deliver out of the box)
- list spacing is still tight compared to Garden's [component](https://garden.zendesk.com/components/lists) – has `::before`/`::after` pseudo been attempted to get the desired 2px?
- the lack of a sticky bottom bar is nearly unusable. Seems like designs are needed to spec out minimum, scrollable editing heights etc.
- are plans in place to fix the `rect` around the code block SVG?

File: src/stories/initialData.js
Comment (m-lai) [reply src/stories/initialData.js:8]: > * have you considered how the current HR treatment throws off vertical rhythm? A `<p>..<p>..<hr>..<p>..<p>` progression results in `16px..20px..20px..16px` of spacing. I'm going to keep asserting that 0 padding on the HR is the right amount (and provides a more general focus solution for "typearound" widgets – which would be very nice to deliver out of the box)

IMO although the 16/20px progression isn't consistent, I felt like it still feels [visually balanced](https://share.getcloudapp.com/mXu1Bq4R). I believe the current HR padding is 4px; I tested 2px with Austin at Allison's suggestion, and we weren't really able to select the HR with affordability to elements below/above the HR. If we went with 0, does this mean the touch affordance would become even worse? If that's the case I'd vote for keeping it as is.

> * list spacing is still tight compared to Garden's [component](https://garden.zendesk.com/components/lists) – has `::before`/`::after` pseudo been attempted to get the desired 2px?

I think the current line-height is the same as <p> text. We should try to match Garden if possible, although I do find the reduced spacing we have for nested lists visually disturbing (lol).

> * the lack of a sticky bottom bar is nearly unusable. Seems like designs are needed to spec out minimum, scrollable editing heights etc.

Agree with this. Can work this out with Austin for pattern specs at least.

> * are plans in place to fix the `rect` around the code block SVG?
This is on our radar but Austin mentioned it's out of scope for this first release.

File: src/theme/_button.css
Comment (austingreendev) [reply src/theme/_button.css:3]: Updated to use themed variable.

File: src/theme/_horizontal-line.css
Comment (austingreendev) [reply src/theme/_horizontal-line.css:3]: Good point. I've disabled this rule by default and remove any specific stylelint overrides.

File: src/theme/_list.css
Comment (austingreendev) [reply src/theme/_list.css:14]: Not that I could find through their theming API. These base64 inline images are used throughout the default theme so I'm not sure it's something we can remove without a full re-skin.

File: src/theme/_variables.css
Comment (austingreendev) [reply src/theme/_variables.css:23]: I was able to get `rrggbbaa` working as a standalone plugin, but providing the alpha in the 8-digit hex makes `simple-var` no longer recognize the theme constant. I also attempted `postcss-env` with `stage: false` to try and isolate the issue, but it seems to be with `color-mod` itself 🤦

File: src/theme/_variables.css
Comment (austingreendev) [reply src/theme/_variables.css:48]: Updated to use themed variable.

File: src/theme/_variables.css
Comment (austingreendev) [reply src/theme/_variables.css:57]: Same issue as above with the 8-digit hex.

File: src/stories/initialData.js
Comment (austingreendev) [reply src/stories/initialData.js]: I've updated the initial data to have a more coherent nested list wording and code-block sizing.

+1 to everything @m-lai said. Having a `Paragraph -> List -> Paragraph` starts to look odd when the line-heights are handled differently. This is also compounded when there is wrapped text content within a list item.

Comment (jzempel) [reply src/stories/initialData.js]: @austingreendev I agree. List spacing cannot be obtained by increasing the `line-height`. But have approaches been explored where `::before/::after` pseudo is used on the `<li>` in order to match the 2px div padding we see with the system? The system has already been considered with other components (i.e. Paragraph) in mind, and we should do everything possible to attempt to match that.

Comment (jzempel) [reply src/stories/initialData.js]: @m-lai I'll ask and then stop pestering, I promise. Why does the HR need touch affordance? I just don't think it does and 0 padding is not only acceptable, but desirable for matching the intended HTML output.

Comment (m-lai) [reply src/stories/initialData.js]: @jzempel I just had a sync with Austin and tried some stuff, the vertical rhythm is now 16px between `<p>...<hr>...<p>` 

>Re: Why does the HR need touch affordance?
The HR is focusable because it's not just for decoration. IMO if it has to be focusable, we might as well make it reasonably easier for users to click, if they wanted to (everything else inside the editor can be clicked on and modified). CKEditor considers it a widget for whatever reason. The typing cursor also takes on the box shadow height so it becomes a 1px blinking dot which looks weird (this adds to it but isn't the main reason why).

File: src/theme/_horizontal-line.css
Comment (jzempel) [inline src/theme/_horizontal-line.css:14]: Since there doesn't appear to be RTL differentiation, would it be better to use cross-browser compatible shorthand?

`margin: 0 calc(var(--ck-spacing-medium) * -1);`

**Review by jzempel — COMMENTED** [review]
There appears to be a pretty strong reliance on `css-bedrock` throughout. If that can't be guaranteed, then items like `background-color` and `line-height` might need to be reconsidered hollistically.

File: src/theme/_horizontal-line.css
Comment (austingreendev) [reply src/theme/_horizontal-line.css:14]: `margin-block` allows the margins to collapse against each other, which is used throughout the other typography elements.

https://caniuse.com/mdn-css_properties_margin-block-start we should be good on browser support since CK doesn't support IE11.

**Review by m-lai — APPROVED** [review]

**Comment by austingreendev** [comment]
@jzempel this ambiguity follows the intended specificity from the lark theme. These editors will be used in a variety of locations, including those that don't have bedrock consistent baselines.

We have applied the relevant bedrock values within the common `_variables` declarations. Are there any specific values that you'd like us to override?

**Review by allisonacs — CHANGES_REQUESTED** [review]
This is feeling really good. A few comments. 

**What's up with this icon?**
image

@austingreendev Why is it out of scope to fix it in this release?

**HR spacing**
I think we're having a classic quandry between mathematical simplicity and visual balance, compounded by the way type generally is written and displayed on the screen, vs how it is perceived. 

image

Ignore the box model for a second. In this image, the red blocks show different measurements from the HR to the type at various points. You can see that it's 20px from the HR to the _baseline_ of the line above, and 20px from the HR to the _cap-height_ of the line below. Those are equal numbers, and from a numbers perspective, creates a predictable system. But because the most common scenario for type displays is sentence case, our eyes don't care that there's an even number on the cap-height, it cares about the x-height (height of lower case x). And those measurements are pretty off. HR to baseline = 20px, but HR to x-height is more like 23-24px. So visually, it's appearing that the type below the HR is further than the type above the HR.

Changing the spacing from the currently implemented (and mathematically pleasing) `16px…16px…16px…16px` to a larger space like `16px…20px…20px…16px` fools our eyes a bit and the extra space makes things feel more even, even if they're not mathematically. If we're going to throw off the vertical rhythm in favor of visual balance, I'd encourage us to commit, and see what something like `16px…20px…18px…16px` looks like. Not saying that has to be how it ends up, but let's look at that vs. 16px uniform vs. 20px uniform, and see how we feel. 

**My thoughts on the HR focus**
Another quandry.
- I think @jzempel's point about inheritance on other focusable widgets is spot on. It would be exceedingly weird for ex. a table to have this vertical spacing between the element and the focus state. Our focus states in Garden are solid, and I'm not sure if visually there's a lot of value in this gap, especially considered under a systemic lens. It also seems like a burden to making fixing the focus styling for other widgets the responsibility of people coding them. Those sorts of things seem like they should work out of the box.
- @m-lai's point that whether we like it or not, CKEditor treats these HRs as widgets, and therefore they should be selectable like all the other widgets is also very valid, and it definitely would be weird to have a single pixel tall blinking dot for the cursor when you're nav-ing around via keyboard.
- Opinion: I think users are going to try to click on the HR. Do we need to optimize for this scenario? I don't know. You can't click to focus a google doc's HR, and when you click around it, the cursor is not a normal height. The interaction does feel off to me. It's still frustrating as a user to perform an action I believe should work, and then it doesn't—even if I know how to overcome that. 
- So, in a perfect world, I think we'd be able to click a reasonably sized click target for the HR (even if it doesn't fully meet WCAG touch target guidelines), but we would use our typical visualization for the focus state. Not sure if that's possible.

**Odds and ends**
_Underlines with headings_
On the subject of line spacing, IMO, the underlines on the headings also feel a little weird. 
image

They're don't have the same spacing as the HRs, but users are going to perceive them to be the same element (that is, they'll think the line and the header are separate elements, and that the header line is actually an HR). Is the line under the headings necessary? I know it's how github does it but… can we establish hierarchy without them? We don't use underlines like this anywhere else in Zendesk.

_Pixel shifting when toggling inline code_
Does it have to happen? What's causing it?

**Comment by jzempel** [comment]
re: https://github.com/zendeskgarden/ckeditor/pull/7#issuecomment-777780242

> Are there any specific values that you'd like us to override?

At a minimum, it seems like we should insist on a with `background-color` rather than allowing the default transparent, since none of the content is designed to work on a dark background.

Screen Shot 2021-02-12 at 8 20 07 AM

**Comment by jzempel** [comment]
re: https://github.com/zendeskgarden/ckeditor/pull/7#pullrequestreview-589190187

> What's up with this icon?

@allisonacs, @m-lai has a task to clean up the SVG (the internal `<rect>` doesn't display correctly in all cases). It doesn't need to hold up this particular PR, but I think we should have it within the example prior to release.

@austingreendev even with that, the following CK lark styling is bogus for us and needs to get knocked out so that the chevron (or other Garden SVG shapes) isn't unintentionally filled:

```css
.ck.ck-icon :not([fill]) {
    fill: currentColor;
}
```

**Comment by jzempel** [comment]
re: https://github.com/zendeskgarden/ckeditor/pull/7#pullrequestreview-589190187

> Pixel shifting when toggling inline code
> Does it have to happen? What's causing it?

@allisonacs getting the component to fit designs was a trick. We have a 13px font size (bump down one, to better match surrounding sans serif) inside of an 18px box – leaving us with an odd number of pixels top/bottom. And all meant to work well within a 20px line-height. 🤔  However, now that we're seeing more uses on the website and here, I'm thinking we might be able to do better. It even seems that CKEditor's `1.95px` hack is rendering better than our `1.5px`.

File: src/theme/_code.css
Comment (jzempel) [inline src/theme/_code.css:4]: `.ck-content pre code` should not be inheriting this padding.

**Comment by jzempel** [comment]
re: https://github.com/zendeskgarden/ckeditor/pull/7#issuecomment-778303843

On close inspection between these two shots, we can see that the application of 3px padding is unpredictable 😩 

- CKEditor thin on top...
Screen Shot 2021-02-12 at 8 51 15 AM

- website thin on bottom...
Screen Shot 2021-02-12 at 8 52 29 AM

File: src/theme/_code.css
Comment (austingreendev) [reply src/theme/_code.css:4]: Updated to override padding.

**Comment by austingreendev** [comment]
@jzempel I've removed their Firefox SVG fill hack, removed inherited code-block padding, and applied a default background-color.

@m-lai let's sync up to try this new vertical spacing rhythm.

**Review by jzempel — APPROVED** [review]
lgtm, pending comments from @allisonacs

**Comment by m-lai** [comment]
> @m-lai let's sync up to try this new vertical spacing rhythm.

Sounds good, let's catch up during our RTE sync when you're back. It was 16/20/16/20 before IIRC which was visually balancing but Allison's typography explainer sheds much more light on reasons why. We can try some of the suggested values and see how we go!

**Comment by allisonacs** [comment]
Capturing the convo between @m-lai and I IRL:

- Discussed the HR spacing issues. Seeing examples with asian characters, and thinking through issues with using images or tables, the decision is to revert back to Mandy's original recommendation of 20px top/bottom spacing on the HR (which is effectively 0 margin added/subtracted, vs. -4px as current). While this creates some differences in vertical rhythm, it provides the best visual balance and is a reasonable approach that makes sense for more languages and content types.
- I was made aware that the HRs are not actually attached to the header styles, but were inserted as part of the example content. These were super confusing for me, as I assumed that the HRs were bundled with the heading style. Recommendation is to use a different example.
- Got the download on the SVG icon
- I think we need to show a max-height on the example, so people get the idea of how it looks and is intended to work. I think the current example, where you have to scroll down to get the toolbar, is disorienting and a bad user experience (so much scrolling into invoke styles).

**Comment by allisonacs** [comment]
> re: [#7 (review)](https://github.com/zendeskgarden/ckeditor/pull/7#pullrequestreview-589190187)
> 
> > Pixel shifting when toggling inline code
> > Does it have to happen? What's causing it?
> 
> @allisonacs getting the component to fit designs was a trick. We have a 13px font size (bump down one, to better match surrounding sans serif) inside of an 18px box – leaving us with an odd number of pixels top/bottom. And all meant to work well within a 20px line-height. 🤔 However, now that we're seeing more uses on the website and here, I'm thinking we might be able to do better. It even seems that CKEditor's `1.95px` padding hack is rendering better than our `1.5px`.

Interesting. Is there an action for us to take here then?

**Review by allisonacs — APPROVED** [review]

**Comment by jzempel** [comment]
> Is there an action for us to take here then?

@allisonacs possibly. We've observed hi-density (half pixel) browser rendering changing over time, so I don't really want to be chasing that. But if we can come up with a robust solution in `react-components` we'll be sure to port the styling over here. It didn't feel important when text was static, but we probably have reason to revisit this now.

**Comment by jzempel** [comment]
@austingreendev if you want to apply the forthcoming fix for the `<code>` pixel shift, it's to reduce the `line-height` by 1px. I haven't explored how to wire up the CSS for an unknown parent `line-height` – something related to `ems` I suppose. Or maybe just fallback to a `line-height: 1`. idk

See https://stackoverflow.com/a/8395743 for details.

**Review by jzempel — CHANGES_REQUESTED** [review]
Please add the following:
- set `line-height: normal;` for `.ck-content code`
- upgrade to `svg-icons` 6.29.0

**Review by jzempel — APPROVED** [review]

**Pull request merged**
